logstash-core/src/main/java/org/logstash/Logstash.java

```diff
@@ -0,0 +1,144 @@
+        config.setOutput(output);
+        config.setError(error);
+        config.setInput(input);
+        script = config.getScriptSource();
+        ruby = Ruby.newInstance(config);
+    }
+
+    @Override
+    public void run() {
+        // @todo: Refactor codebase to not rely on global constant for Ruby Runtime
+        if (RubyUtil.RUBY != ruby) {
+            throw new IllegalStateException(
+                "More than one JRuby Runtime detected in the current JVM!"
+            );
+        }
+        try {
+            Thread.currentThread().setContextClassLoader(ruby.getJRubyClassLoader());
+            ruby.runFromMain(script, config.displayedFileName());
+        } catch (final RaiseException ex) {
+            final RubyException rexep = ex.getException();
+            if (ruby.getSystemExit().isInstance(rexep)) {
+                final IRubyObject status =
+                    rexep.callMethod(ruby.getCurrentContext(), "status");
+                if (status != null && !status.isNil() && RubyNumeric.fix2int(status) != 0) {
+                    throw new IllegalStateException(ex);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        ruby.tearDown(false);
+        script.close();
+    }
+
+    /**
+     * Sets up the correct {@link RubyInstanceConfig} for a given Logstash installation and set of
+     * CLI arguments.
+     * @param home Logstash Root Path
+     * @param args Commandline Arguments Passed to Logstash
+     * @return RubyInstanceConfig
+     */
+    private static RubyInstanceConfig buildConfig(final Path home, final String[] args) {
+        final String[] arguments = new String[args.length + 2];
+        System.arraycopy(args, 0, arguments, 2, args.length);
+        arguments[0] = safePath(home, "lib", "bootstrap", "environment.rb");
+        arguments[1] = safePath(home, "logstash-core", "lib", "logstash", "runner.rb");
+        final RubyInstanceConfig config = new RubyInstanceConfig();
+        config.processArguments(arguments);
+        return config;
+    }
+
+    /**
+     * Builds the correct path for a file under the given Logstash root and defined by its sub path
+     * elements relative to the Logstash root.
+     * Ensures that the file exists and throws an exception of it's missing.
+     * This is done to avoid hard to interpret errors thrown by JRuby that could result from missing
+     * Ruby bootstrap scripts.
+     * @param home Logstash Root Path
+     * @param subs Path elements relative to {@code home}
+     * @return Absolute Path a File under the Logstash Root.
+     */
+    private static String safePath(final Path home, final String... subs) {
+        Path resolved = home;
+        for (final String element : subs) {
+            resolved = resolved.resolve(element);
+        }
+        if (!resolved.toFile().exists()) {
+            throw new IllegalArgumentException(String.format("Missing: %s.", resolved));
```

Comment:
Can we make this more actionable, or would a user never expect to see this?

Reply:
@jordansissel jup same here => user should never ever see this except for when there's a bug :)